torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py

```diff
@@ -75,10 +75,193 @@ def powerSGD_hook(
     bucket,
 ) -> torch.futures.Future:
     """
-    This DDP communication hook implements a simplified PowerSGD gradient compression
+    This DDP communication hook implements the original PowerSGD gradient compression
     algorithm described in https://arxiv.org/abs/1905.13727.
     Once gradient tensors are aggregated across all workers, this hook applies
     compression as follows:
+    1) Views the input flattened 1D gradient tensor as two groups of per-parameter tensors:
+    high-rank tensors and vector-like rank-1 tensors (for biases).
+    2) Handles rank-1 tensors by allreducing them without compression:
```

Comment:
Does the paper do this shortcut for allreduce 1d tensors directly or is this something that we've added ourselves?

Reply:
Same as the paper implementation. The input of the compression is a matrix, so it does not work on 1D tensors.